packages/layers/src/xr-layer/xr-layer.js

```diff
@@ -122,11 +128,13 @@ const XRLayer = class extends Layer {
     // might be created, this solves the performance issue of always adding new
     // hook functions.
     // See https://github.com/kylebarron/deck.gl-raster/blob/2eb91626f0836558f0be4cd201ea18980d7f7f2d/src/deckgl/raster-layer/raster-layer.js#L21-L40
-    if (!programManager._hookFunctions.includes(mutateStr)) {
-      programManager.addShaderHook(mutateStr);
+    // Note: _hookFunctions is private, not sure if there's an appropriate way to check if a hook function is already added.
+    // it may be better to add these hooks somewhere else rather than in initializeState of a layer?
```

Comment:
I'm pretty sure I yanked this from the deck.gl internals.  In case you couldn't tell, a lot of this stuff comes from me just looking at what they do 😃

Reply:
So it does seem like this might be gone: https://github.com/search?q=repo%3Avisgl%2Fdeck.gl+_hookFunctions&type=code  I'll need to look into how they add shader hooks.  In general, I have found (and you seem to have as well) that while the docs make clear what things do, they don't exactly make clear how to do them and then you end up digging around.

Reply:
`_hookFunctions` seemed to still be there when I was stepping through in the debugger - interesting that search comes up empty, but of course we'd hit a very obvious error otherwise. I don't think it's causing a problem being the way it is, was flagged (along with plenty of other things of course) by TypeScript and at some point it could be changed, but I'm pretty there's nothing specific that relates to any problems we're experiencing. Hypothetically we could re-arrange that to be a bit more respectful of the intended use.

Incidentally, a hidden github shortcut I find quite useful is that if you press '.' it converts the browser tab into a vscode instance for the project. I often find that more useful for navigating and searching code than the normal web interface.

Reply:
I think the reason it doesn't appear in deck.gl is that it's a luma.gl internal thing.

Reply:
Also, I've found that temporarily renaming to e.g. `xr-layers.ts` means that I can more often do things like 'go to definition' (cmd+click) to get to the deck/luma implementation of things.

Reply:
Weird though that it doesnt come up as I might expect deck.gl to do this at some point, so maybe there's a new way to register hooks.  I was just thinking, and I believe that I was at one point _told_ to do this.  In any case, glad it works and I can ask the devs as well.

Reply:
https://github.com/visgl/deck.gl/blob/a74533027c9607e3c180859eee73086027350d6b/docs/upgrade-guide.md#shader-assembly

> Custom layers are no longer expected to call assembleShaders directly. Instead, the new Model class from luma.gl v4 will take shaders and the modules they are using as parameters and assemble shaders automatically.

It seems to me that we're calling `Model` as the docs say we should, the shaders are being assembled and attached... just not compiled. Actually, I think perhaps that they *are* compiled, but the `compilationStatus` flag is set to `"pending"` and...

From luma.gl [`webgl-shader.ts`](https://github.com/visgl/luma.gl/blob/d8db53635150af9007b3a0498fe778d58996d1f2/modules/webgl/src/adapter/resources/webgl-shader.ts#L4)

```ts
  /** Compile a shader and get compilation status */
  protected async _compile(source: string): Promise<void> {
    const addGLSLVersion = (source: string) =>
      source.startsWith('#version ') ? source : `#version 100\n${source}`;
    source = addGLSLVersion(source);

    const {gl} = this.device;
    gl.shaderSource(this.handle, source);
    gl.compileShader(this.handle);

    // For performance reasons, avoid checking shader compilation errors on production
    if (log.level === 0) {
      this.compilationStatus = 'pending';
      return;
    }
    ....
```

That's the path it's going down... and it doesn't seem to call `_getCompilationStatus()` at any point after that, which would seem like the mechanism for it to update that `compilationStatus`...

So I don't know if this is just a luma bug, or what - clearly other shaders are being compiled and working, and I don't really see how this part of the logic would be different for the xr-layer shaders than e.g. scale bars which also go through this code... but it seems like a decent lead, anyway.

Reply:
🎉

I've added this in `xr-layer.js`

```js
import { log } from '@luma.gl/core';
// force lumagl webgl-shader to take a code-path that actually compiles shaders & gives us an error...
log.setLevel(1);
```

Now it goes down a different path, and gives actual shader compilation error output.

Oh, happy day.

Reply:
OK... I think I've made relevant shader edits. Now I get

```
deck: drawing XRLayer({id: 'image-sub-layer-0,131,175,0-Background-Image-TiffPixelSource-#detail#'}) to screen: No value for binding channel0 in image-sub-layer-0,131,175,0-Background-Image-TiffPixelSource-#detail#-cached Error: No value for binding channel0 in image-sub-layer-0,131,175,0-Background-Image-TiffPixelSource-#detail#-cached
    at WEBGLRenderPipeline._applyBindings (webgl-render-pipeline.js:314:23)
    at WEBGLRenderPipeline.draw (webgl-render-pipeline.js:167:14)
    at _Model.draw (model.js:256:41)
    at XRLayer.draw (xr-layer.js:278:13)
    at layer.js:851:22
    at withGLParameters (with-parameters.js:17:16)
    at _WebGLDevice.withParametersWebGL (webgl-device.js:269:16)
    at XRLayer._drawLayer (layer.js:845:28)
    at DrawLayersPass._drawLayersInViewport (layers-pass.js:153:27)
    at DrawLayersPass._drawLayers (layers-pass.js:54:36)
```